### PR TITLE
Add more number converter tests

### LINKED ISSUE #36649: BaseNumberConverter.CanConvertTo(null) throws NRE should return false

### LINKED ISSUE #36659: BaseNumberConverter ConvertFrom for decimal/double/float differs from .NET Framework and is buggy anyway
From .NET Framework reference source:
```cs
public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value) {
    if (value is string) {
        string text = ((string)value).Trim();

        try {
            if (AllowHex && text[0] == '#') {
                return FromString(text.Substring(1), 16);
            }
            else if (AllowHex && text.StartsWith("0x") 
                        || text.StartsWith("0X")
                        || text.StartsWith("&h")
                        || text.StartsWith("&H")) {
                return FromString(text.Substring(2), 16);
            }
            else {
            }
            ...
        }
    }
}
```

In .NET Core:

```cs
public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
{
    if (value is string text)
    {
        text = text.Trim();

        try
        {
            if (AllowHex && text[0] == '#')
            {
                return FromString(text.Substring(1), 16);
            }
            else if (AllowHex && text.StartsWith("0x", StringComparison.OrdinalIgnoreCase)
                        || text.StartsWith("&h", StringComparison.OrdinalIgnoreCase))
            {
                return FromString(text.Substring(2), 16);
            }
    }
}
```

This causes behavioural differences in the following situations, applicable to `DecimalConverter`, `DoubleConverter` and `SingleConverter`, because all these set `AllowHex` to false.

- `converter.ConvertFrom("#2")`: throws in .NET Framework; throws in .NET Core
- `converter.ConvertFrom("0x2")`: throws in .NET Framework; throws in .NET Core
- `converter.ConvertFrom("0X2")`: *does not* throw in .NET Framework; throws in .NET Core
- `converter.ConvertFrom("&h2")`: *does not* throw in .NET Framework; *does not* throw in .NET Core
- `converter.ConvertFrom("&H2")`: *does not* throw in .NET Framework; *does not* throw in .NET Core

There are several bugs, both in .NET Core and in .NET Framework due to incorrect order of precedence between the && and the || operators.

The intention of the code is to disallow (throw) the following:
- `converter.ConvertFrom("#2")`
- `converter.ConvertFrom("0x2")`
- `converter.ConvertFrom("0X2")`
- `converter.ConvertFrom("&h2")`
- `converter.ConvertFrom("&H2")`

This is because all these representations above are hex, but `AllowHex` only applies to the first condition in the following list of `... || ...` due to the order of precedence. Changed bracketing would fix this

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #36649
Fixes #36659

### CODE REVIEW

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs [hunk 1]
```diff
@@ -45,16 +42,20 @@ public abstract class BaseNumberConverter : TypeConverter
         /// </summary>
         public override bool CanConvertFrom(ITypeDescriptorContext context, Type sourceType)
         {
-            return sourceType == typeof(string) || base.CanConvertFrom(context, sourceType);
+            if (sourceType == typeof(string))
+            {
+                return true;
+            }
+
+            return base.CanConvertFrom(context, sourceType);
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs]: Why did you change this code?  The previous style seems perfectly fine.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs [hunk 2]
```diff
@@ -111,13 +115,18 @@ public override object ConvertTo(ITypeDescriptorContext context, CultureInfo cul
...
 
         public override bool CanConvertTo(ITypeDescriptorContext context, Type destinationType)
         {
-            return base.CanConvertTo(context, destinationType) || destinationType.IsPrimitive;
+            if (destinationType != null && destinationType.IsPrimitive)
+            {
+                return true;
+            }
+
+            return base.CanConvertTo(context, destinationType);
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs]: Why did you change this?

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs [hunk 3]
```diff
@@ -31,7 +32,12 @@ public class DecimalConverter : BaseNumberConverter
         /// </summary>
         public override bool CanConvertTo(ITypeDescriptorContext context, Type destinationType)
         {
-            return destinationType == typeof(InstanceDescriptor) || base.CanConvertTo(context, destinationType);
+            if (destinationType == typeof(InstanceDescriptor))
+            {
+                return true;
+            }
+
+            return base.CanConvertTo(context, destinationType);
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs]: Why did you change this?

I don't understand what these stylistic changes have to do with adding more tests as the PR is titled.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs [hunk 4]
```diff
@@ -43,34 +49,20 @@ public override bool CanConvertTo(ITypeDescriptorContext context, Type destinati
...
-                }
-            }
-            
             return base.ConvertTo(context, culture, value, destinationType);
         }
 
         /// <summary>
         /// Convert the given value to a string using the given radix
         /// </summary>
+        [ExcludeFromCodeCoverage] // Not called as AllowHex is false.
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs]: Please remove this.  If we have code that's not exercised, it should show up as such in a code coverage report.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DoubleConverter.cs [hunk 5]
```diff
@@ -25,6 +26,7 @@ public class DoubleConverter : BaseNumberConverter
         /// <summary>
         /// Convert the given value to a string using the given radix
         /// </summary>
+        [ExcludeFromCodeCoverage] // Not called as AllowHex is false.
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DoubleConverter.cs]: Ditto for all of these.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs (hunk 4 above)
Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs]: But it genuinely can’t be called. Decimal/double/single converter overrides `AllowHex` to return false. Therefore the convert to string overload will provably never be called because we never do hex parsing as a result. So this exclusion attribute can be justified.

Comment (stephentoub) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs]: > But it genuinely can’t be called.

I understand.  But our goal isn't to get to 100% code coverage just so we can have a 100% number.  The only reason we look at code coverage is as an indicator to help drive better testing; hiding code from the report like this only serves to artificially increase the number and prevents us from seeing that we have code that's never getting executed.  If it's actually not reachable, than the fix (not for this PR) is to delete the dead code, not hide it from a code coverage report so that no one in the future will be able to see that it's not being executed.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs (hunk 2 above)
Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs]: This fixes a `NullReferenceException` if `destinationType` is null

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs (hunk 1 above)
Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/BaseNumberConverter.cs]: Well the line of code didnt fit on my screen so i wanted to format it. Reverted

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs (hunk 4 above)
Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs]: Very well, removed

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs (hunk 3 above)
Comment (hughbe) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/DecimalConverter.cs]: reverted

**Comment by hughbe** [comment]
Rebased :)

File: src/System.ComponentModel.TypeConverter/tests/TypeConverterTestBase.cs [hunk 6]
```diff
@@ -78,6 +79,18 @@ public void ConvertTo_NullDestinationType_ThrowsArgumentNullException()
...
         }
 
+        [Fact]
+        public void CanConvertTo_NullDestinationType_ReturnsFalse()
+        {
+            try
+            {
+                Assert.False(Converter.CanConvertTo(null));
+            }
+            catch (NullReferenceException) when (PlatformDetection.IsFullFramework)
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/tests/TypeConverterTestBase.cs]: We should instead just skip the test on netfx.

File: src/System.ComponentModel.TypeConverter/tests/TypeConverterTestBase.cs [hunk 7]
```diff
@@ -153,6 +166,20 @@ public void GetStandardValuesExclusive_Invoke_ReturnsExpected()
             Assert.Equal(StandardValuesExclusive, converter.GetStandardValuesExclusive());
         }
 
+        private static void AssertEqual(object expected, object actual)
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/tests/TypeConverterTestBase.cs]: We should name this differently; otherwise it looks like a typo at the call site, e.g. AssertInstanceDescriptorsEqual, or something like that.

**Comment by stephentoub** [comment]
Thanks.
